_stbt/keyboard.py

```diff
@@ -88,70 +82,482 @@ class Keyboard(object):
+        control.
+
+        :param source: The starting key. This can be a Key object returned from
+            `add_key` or `find_key`; or it can be a dict that contains one or
+            more of "name", "text", "region", and "mode" (as many as are needed
+            to uniquely identify the key using `find_key`). For convenience, a
+            single string is treated as "name" (but this may not be enough to
+            uniquely identify the key if your keyboard has multiple modes).
+
+        :param target: The key you'll land on after pressing the button
+            on the remote control. This accepts the same types as ``source``.
+
+        :param str keypress: The name of the key you need to press on the
+            remote control, for example "KEY_RIGHT".
 
-    # pylint:disable=fixme
-    # TODO: case sensitive keyboards
-    #   Caps lock can be supported with a graph like this:
-    #       A CAPSLOCK_OFF KEY_LEFT
-    #       CAPSLOCK_OFF CAPSLOCK_ON KEY_OK
-    #       CAPSLOCK_ON a KEY_RIGHT
-    #       a q KEY_UP
-    #       <etc>
-    #   (Other mode changes like ABC -> 123!@# can be supported in the same
-    #   way.)
-    #
-    #   I don't know how best to support SHIFT (temporary mode change):
-    #   - In shifted state KEY_OK will go from "A" to "a". We'd want to use
-    #     press_and_wait before we check the new state. But...
-    #   - In non-shifted state KEY_OK just enters the letter; the screen doesn't
-    #     change otherwise. Currently we don't use press_and_wait here because
-    #     typically the text-box where the text appears is masked out (because
-    #     some UIs have a blinking cursor there) and there is no change anywhere
-    #     else on the screen.
-    #
-    # TODO: Check that KEY_OK adds to the search text? With a property on the
-    #   page object? OCR might be too unreliable for incomplete words. We don't
-    #   use press_and_wait because the text-box might be masked out (some UIs
-    #   have a blinking cursor there).
+        :param str mode: Optional keyboard mode that applies to both ``source``
+            and ``target``. For example, the two following calls are the same::
+
+                add_transition("c", " ", "KEY_DOWN", mode="lowercase")
+
+                add_transition({"name": "c", "mode": "lowercase"},
+                               {"name": " ", "mode": "lowercase"},
+                               "KEY_DOWN")
+
+        :param bool symmetrical: By default, if the keypress is "KEY_LEFT",
+            "KEY_RIGHT", "KEY_UP", or "KEY_DOWN", this will automatically add
+            the opposite transition. For example, if you call
+            ``add_transition("a", "b", "KEY_RIGHT")`` this will also add the
+            transition ``("b", "a", "KEY_LEFT)"``. Set this parameter to False
+            to disable this behaviour. This parameter has no effect if
+            ``keypress`` is not one of the 4 directional keys.
```

Comment:
I've raised #696 which makes a small change: Moved `SYMMETRICAL_KEYS` from a (private) module constant to a (public) instance member of `Keyboard`. So user code can add/remove keys from it:

    kb = stbt.Keyboard()
    kb.symmetrical_keys["KEY_1"] = "KEY_2"
    kb.symmetrical_keys["KEY_2"] = "KEY_1"

It doesn't solve the backwards compatibility problem, but it gives us a tiny bit more flexibility without having to make a new Stb-tester release, so why not. You could even set it to an empty dict if you want to disable this behaviour but you don't want to have to specify `symmetrical=False` every time you call `add_transition`.

Reply:
> I wonder if we should add a separate method for adding symmetrical keys.

What I meant by this is we should add a separate method for adding symmetrical **transitions**, rather than keys.